src/services/microservices-service.js

```diff
@@ -0,0 +1,196 @@
+ *  * http://www.eclipse.org/legal/epl-2.0
+ *  *
+ *  * SPDX-License-Identifier: EPL-2.0
+ *  *******************************************************************************
+ *
+ */
+
+const TransactionDecorator = require('../decorators/transaction-decorator');
+const MicroserviceManager = require('../sequelize/managers/microservice-manager');
+const MicroservicePortManager = require('../sequelize/managers/microservice-port-manager');
+const VolumeMappingManager = require('../sequelize/managers/volume-mapping-manager');
+const IOFogService = require('../services/iofog-service');
+const FlowService = require('../services/flow-service');
+const AppHelper = require('../helpers/app-helper');
+const Errors = require('../helpers/errors');
+const Validation = require('../schemas/index');
+
+const _getMicroserviceByFlow = async function (flowId, user, transaction) {
+  await FlowService.getFlow(flowId, user, false, transaction);
+
+  const microservice = {
+    flowId: flowId
+  };
+
+  return await MicroserviceManager.findAllWithDependencies(microservice, {}, transaction)
+};
+
+const _getMicroservice = async function (microserviceUuid, user, transaction) {
+  const microservice = await MicroserviceManager.findOneWithDependencies({
+    uuid: microserviceUuid
+  },
+  {}, transaction);
+
+  await FlowService.getFlow(microservice.flowId, user, transaction);
+
+  await IOFogService.getFogWithTransaction({
+    uuid: microservice.iofogUuid
+  }, user, transaction);
+
+  return microservice;
+};
+
+const _createMicroserviceOnFog = async function (microserviceData, user, transaction) {
+  await Validation.validate(microserviceData, Validation.schemas.microserviceCreate);
+
+  const microservice = await _createMicroservice(microserviceData, user, transaction);
+
+  if (microserviceData.port) {
+    await _createMicroservicePort(microserviceData, microservice.uuid, transaction);
+  }
+  if (microserviceData.volumeMappings) {
+    await _createVolumeMappings(microserviceData, microservice.uuid, transaction);
+  }
+  // TODO: create routes
+
+  return {
+    uuid: microservice.uuid
+  }
+};
+
+const _createMicroservice = async function (microserviceData, user, transaction) {
+  const microserviceToCreate = {
+    uuid: AppHelper.generateRandomString(32),
+    name: microserviceData.name,
+    config: microserviceData.config,
+    isNetwork: microserviceData.isNetwork,
+    needUpdate: microserviceData.needUpdate,
+    rebuild: microserviceData.rebuild,
+    catalogItemId: microserviceData.catalogItemId,
+    flowId: microserviceData.flowId,
+    iofogUuid: microserviceData.ioFogNodeId,
+    rootHostAccess: microserviceData.rootHostAccess,
+    logSize: microserviceData.logLimit,
+  };
+
+  const microserviceDataCreate = AppHelper.deleteUndefinedFields(microserviceToCreate);
+
+  if (microserviceDataCreate.flowId) {
+    await FlowService.getFlow(microserviceDataCreate.flowId, user, transaction);
```

Comment:
Do we actually need to do this in order to check if flow exists?
I think we can just try to create microservice and if there is no such flow id you will get 500 error with constraint error in stack trace.

Reply:
it's check if microservice belongs to user via flow

Reply:
looks strange. Why then don't you check if microservice belongs to user via catalog item?
I think since microservice object is dependent on 3 other objects it's better to create separate lightweight query that checks all that 3 dependencies simaltaniously